devito/function.py

```diff
@@ -259,43 +275,137 @@ def _indices(cls, **kwargs):
             dimensions = grid.dimensions
         return dimensions
 
+    def _allocate_memory(func):
+        """Allocate memory as a :class:`Data`."""
+        def wrapper(self):
+            if self._data is None:
+                debug("Allocating memory for %s (%s)" % (self.name, self.shape))
+                self._data = Data(self.shape, self.indices, self.space_order, self.dtype)
+                if self._first_touch:
+                    first_touch(self)
+                else:
+                    self.data.fill(0)
+            return func(self)
+        return wrapper
+
     @property
-    def shape_data(self):
+    def _offset_domain(self):
         """
-        Full allocated shape of the data associated with this :class:`Function`.
+        The number of grid points between the first allocated element
+        (possibly in the halo/padding region) and the first domain element,
+        for each dimension.
         """
-        return tuple(i - s for i, s in zip(self.shape_domain, self.staggered))
+        return tuple(np.add(self._halo, self._padding))
+
+    @property
+    def _offset_halo(self):
+        """
+        The number of grid points between the first allocated element
+        (possibly in the halo/padding region) and the first domain element,
```

Comment:
If I understand this correctly, this should read "the first halo element", right?

Reply:
uops, yes, __doc__ needs fixing; I'll do that in #462 to avoid another Travis round